pkg/cloudprovider/providers/aws/aws_loadbalancer.go

```diff
@@ -62,6 +81,773 @@ func getLoadBalancerAdditionalTags(annotations map[string]string) map[string]str
+							ListenerArn: listener.ListenerArn,
+							Port:        aws.Int64(frontendPort),
+							Protocol:    aws.String("TCP"),
+							DefaultActions: []*elbv2.Action{{
+								TargetGroupArn: targetGroup.TargetGroupArn,
+								Type:           aws.String("forward"),
+							}},
+						})
+						if err != nil {
+							glog.Warning("Unable to update load balancer listener")
+							return nil, err
+						}
+
+						// Delete old target group
+						_, err = c.elbv2.DeleteTargetGroup(&elbv2.DeleteTargetGroupInput{
+							TargetGroupArn: listener.DefaultActions[0].TargetGroupArn,
+						})
+						if err != nil {
+							glog.Warning("Unable to delete old target group")
+							return nil, err
+						}
+
+					} else {
+						// Run ensureTargetGroup to make sure instances in service are up-to-date
+						targetName := createTargetName(namespacedName, frontendPort, nodePort)
+						_, err = c.ensureTargetGroup(
+							targetGroup,
+							mapping,
+							instanceIDs,
+							targetName,
+							*loadBalancer.VpcId,
+						)
+						if err != nil {
+							return nil, err
+						}
+					}
+					dirty = true
+					continue
+				}
+
+				// Additions
+				_, targetGroupArn, err := c.createListenerV2(loadBalancer.LoadBalancerArn, mapping, namespacedName, instanceIDs, *loadBalancer.VpcId)
+				if err != nil {
+					return nil, err
+				}
+				addTagsInput.ResourceArns = append(addTagsInput.ResourceArns, targetGroupArn)
+				dirty = true
+			}
+
+			frontEndPorts := map[int64]bool{}
+			for i := range mappings {
+				frontEndPorts[mappings[i].FrontendPort] = true
+			}
+
+			// handle deletions
+			for port, listener := range actual {
+				if _, ok := frontEndPorts[port]; !ok {
+					err := c.deleteListenerV2(listener)
+					if err != nil {
+						return nil, err
+					}
+					dirty = true
+				}
+			}
+
+			// Add tags to new targets
+			for k, v := range tags {
+				addTagsInput.Tags = append(addTagsInput.Tags, &elbv2.Tag{
+					Key: aws.String(k), Value: aws.String(v),
+				})
+			}
+			if len(addTagsInput.ResourceArns) > 0 && len(addTagsInput.Tags) > 0 {
+				_, err = c.elbv2.AddTags(addTagsInput)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+
+		// Subnets cannot be modified on NLBs
+		if dirty {
+			loadBalancers, err := c.elbv2.DescribeLoadBalancers(
+				&elbv2.DescribeLoadBalancersInput{
+					LoadBalancerArns: []*string{
+						loadBalancer.LoadBalancerArn,
+					},
+				},
+			)
+			if err != nil {
+				glog.Warning("Unable to retrieve load balancer after update")
+				return nil, err
+			}
+			loadBalancer = loadBalancers.LoadBalancers[0]
+		}
+	}
+	return loadBalancer, nil
+}
+
+// create a valid target group name - ensure name is not over 32 characters
+func createTargetName(namespacedName types.NamespacedName, frontendPort, nodePort int64) string {
+	sha := fmt.Sprintf("%x", sha1.Sum([]byte(namespacedName.String())))[:20]
+	return fmt.Sprintf("%s-%d-%d", sha, frontendPort, nodePort)
```

Comment:
Is this visible to external users (e.g. in DNS somewhere)?

Because if not, users hate these random names... we could do the name and a portion of the hash maybe.

Also, in the past we've had problems because some names e.g. can't start with a number.  Any rules like that here?

Reply:
I get no problem creating groups that start with numbers: what about something like

`k8s-tg-<13-char-hash>-<frontend-port>-<backend port>`?

The target groups are still tagged with the k8s service and cluster name